Add optional clustering report template (#207)
AlexsLemonade/scpca-downstream-analyses: optional-clustering-analysis/clustering-report-template.Rmd

---
params:
  library: Library

>   processed_sce: "../results/scpca_processed_data/SCPCS000216/SCPCL000290_miQC_processed_sce.rds"

Reviewer: ```suggestion
  processed_sce: "../results/scpca_processed_data/SCPCS000216/SCPCL000290_miQC_processed_sce_clustering.rds"
```
I think you actually want the input to this to be the output sce object from the clustering script. https://github.com/AlexsLemonade/scpca-downstream-analyses/blob/2faee9afe5d8ac290ae3546653519b5c2fd735ca/optional-clustering-analysis/clustering-calculations.R#L202-L205

  stats_dir: "../results/scpca_processed_data/SCPCS000216/clustering_stats"
  cluster_type: "louvain"
  nearest_neighbors_range: 5:25
  nearest_neighbors_increment: 5
  project_root: NULL

>   mito_file: "reference-files/Homo_sapiens.GRCh38.104.mitogenes.txt"

Reviewer: ```suggestion
```
You can remove this. You shouldn't need the mito file for this notebook.

library(ggpubr)
library(scater)
```

```{r}
# Read in SCE object
processed_sce <- read_rds(file.path(params$processed_sce))
```

## Plots

## Clustering results for `r params$library`

Below is a UMAP plot where each cell has been colored based on the cell's cluster assignment. 
Here clustering was performed using `r params$cluster_type` clustering with a nearest neighbors value of `r params$nearest_neighbors`.

```{r}
# If param$nearest_neighbors_range is a single value, set nearest_neighbors_increment to NULL
if(length(params$nearest_neighbors_range) == 1) {
  nearest_neighbors_increment <- 1
}
# If a nearest neighbors increment exists, then create a sequence for clustering 
if(!(params$nearest_neighbors_increment == 1)){
  nn_range <- seq(min(params$nearest_neighbors_range), 
                  max(params$nearest_neighbors_range), 
                  params$nearest_neighbors_increment) 
  # If no nearest neighbors increment has been input then the provided range is 
  # directly used for clustering 
} else {
  nn_range <- params$nearest_neighbors_range
}

# Grab column name with clustering results
cluster_column_names <- paste(params$cluster_type, nn_range, sep = "_")

> existing_columns <- intersect(cluster_column_names, colnames(colData(processed_sce)))

Reviewer: I would move this all to a separate chunk above the UMAP section and in that section you want to also perform a check that `length(existing_columns)` is not 0. Otherwise clustering hasn't been performed using the given parameters and the report can't be created. 

Additionally I think it might be time to break out the creation of the `nn_range` and the checks around that into its own function. The function would take as input the `nearest_neighbors_range` and the `nearest_neighbors_increment`, do the steps in line 79-92 and then return the `nn_range`.

# Plot
plotReducedDim(processed_sce, dimred = "UMAP", colour_by = existing_columns) +
  theme_bw() +
  labs(
    caption = paste0(
      stringr::str_to_title(params$cluster_type),
      " clustering with nearest neighbors value ",
      params$nearest_neighbors_range
    )
  ) +
  guides(col = guide_legend("Cluster assignment")) +
  theme(text = element_text(size = 14),
        plot.caption = element_text(hjust = 0.5))
```

## Plot cluster validity stats

```{r fig.height = 9.5}
# Grab the cluster validity stats data frame
all_validity_stats_df <- readr::read_tsv(file.path(
  params$stats_dir,
  paste0(params$library, "_clustering_all_validity_stats.tsv")

> ))

Reviewer: After reading this in, I would add a check that makes sure that data corresponding to the `existing_clusters` is present and if it's not there I would throw an error. I also would probably subset the dataframe to those `existing_clusters` cluster names. This should apply to all of the dataframes.

# Grab column name with clustering results
cluster_column_names <- paste(params$cluster_type, nn_range, sep = "_")
existing_columns <- intersect(cluster_column_names, colnames(colData(processed_sce)))

# Check length of `existing_columns`
if(length(existing_columns) == 0){
  stop("There are no existing clustering results, therefore a report cannot be generated.")
}

# Set number of columns for plotting and use this to define figure heights
num_col <- 2
figure_height <- max(ceiling(length(existing_columns)/num_col) * 5, 4)

knitr::opts_chunk$set(
  fig.height = figure_height
)
```

```{r}
# Plot UMAP coordinates
plot_list <- existing_columns %>%
  purrr::map(
    ~ plotReducedDim(processed_sce, dimred = "UMAP", colour_by = .x) +
      theme_bw() +
      theme(text = element_text(size = 8))
  )

cowplot::plot_grid(plotlist = plot_list, ncol = num_col)
```

## Plot cluster validity stats

```{r}
# Grab the cluster validity stats data frame
all_validity_stats_df <- readr::read_tsv(file.path(
  params$stats_dir,
  paste0(params$library, "_clustering_all_validity_stats.tsv")
))

# Check that the clustering results exist

> if(!(any(existing_columns %in% all_validity_stats_df$cluster_names_column))){
>   stop("The clustering results cannot be found in the all validity stats file.")
> } else {
>   all_validity_stats_df <- all_validity_stats_df %>%
>     dplyr::filter(cluster_names_column %in% existing_columns)
> }

Reviewer: ```suggestion
stats_column_check <- function(existing_columns, df_column){
  missing_columns <- setdiff(existing_columns, df_column)
  if(length(missing_columns) != 0){
    missing_columns_print <- paste(missing_columns, collapse = ", ")
    stop(glue::glue("The clustering statistics cannot be found for {missing_columns_print}")
    }
  }
stats_column_check(existing_columns, all_validity_stats_df$cluster_names_column)

all_validity_stats_df <- all_validity_stats_df %>%
    dplyr::filter(cluster_names_column %in% existing_columns)
```

So the first thing I would do is create a small subfunction that does the checking for the columns in each of the columns of the dataframes for you so you don't have to repeat the check and error printing everytime. I wrote an example here but the spacing might be a little off so make sure you double check that. 

I also would move these checks and reading in each of the files to the beginning of the notebook after the check for the existing columns in the SCE object. If there are no stats we want to know before we make any UMAPs. 

The last thing I would do is make sure we are checking that all of the clustering results for the whole range of parameters (e.g. all `existing_columns`) have associated statistics. Using `any` will just check that at least one of them is present.

---
params:
  library: Library
  processed_sce: "../results/scpca_processed_data/SCPCS000216/SCPCL000290_processed_sce_clustering.rds"

>   stats_dir: "../results/scpca_processed_data/SCPCS000216/clustering_stats"

Reviewer 2: There are further issues here - I went to rerun the Snakefile in order to generate input result files for this notebook. This failed because of how the argument `nearest_neighbors_increment` is being used in `define_nn()`. The `if` statement: `if !(nearest_neighbors_increment == 1)` cannot actually be evaluated if this argument is `NULL`. I've made a suggestion at that line that should do the trick to revive the Snakefile.

>   stats_dir: "../results/scpca_processed_data/SCPCS000216/clustering_stats"

Author: Filed an issue to address that README update - https://github.com/AlexsLemonade/scpca-downstream-analyses/issues/211

  cluster_type: "louvain"
  nearest_neighbors_range: 5:25
  nearest_neighbors_increment: 5
  project_root: NULL
  date: !r Sys.Date()

title: "`r glue::glue('Clustering analysis report for {params$library}')`"
author: "CCDL"
date: "`r params$date`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_folding: hide
---

This notebook will look at additional clustering parameters on the expression of

> the genes in a single sample of the dataset of interest, from an unbiased approach.

Reviewer: Just want to note that we have a separate issue for adding documentation and explanation into the notebook #206 since we want to expand on this a lot more in general which is going to take another round of review.

> the genes in a single sample of the dataset of interest, from an unbiased approach.

Author: Linking the PR here for reference: https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/210

    dplyr::filter(cluster_names_column %in% existing_columns)

summary_validity_stats_df <- summary_validity_stats_df %>%
    dplyr::filter(cluster_names_column %in% existing_columns)

summary_stability_stats_df <- summary_stability_stats_df %>%
    dplyr::filter(cluster_names_column %in% existing_columns)
```

## Plots

## UMAP plot results for `r params$library`

Below is a UMAP plot where each cell has been colored based on the cell's cluster assignment. 
Here clustering was performed using `r params$cluster_type` clustering.

```{r}
# Set number of columns for plotting and use this to define figure heights
num_col <- 2
figure_height <- max(ceiling(length(existing_columns)/num_col) * 5, 4)

knitr::opts_chunk$set(
  fig.height = figure_height
)
```

```{r}
# Plot UMAP coordinates
plot_list <- existing_columns %>%
  purrr::map(
    ~ plotReducedDim(processed_sce, dimred = "UMAP",colour_by = .x, point_size = 0.1, point_alpha = 0.3) +
      theme_bw() +
      guides(color = guide_legend(override.aes = list(size = 1)))
  )

cowplot::plot_grid(plotlist = plot_list, ncol = num_col)
```

## Plot cluster validity stats

#### Purity plots

```{r}
# Plot individual cluster purity stats

> purity_plots <- plot_cluster_purity(all_validity_stats_df, num_col)

Reviewer: Something about these plots looks off to me. The legend here is really big and almost the center of attention while there are two squished columns with data. I almost think for this one we need three columns...
